matchers/receive_matcher.go

```diff
@@ -29,15 +30,38 @@ func (matcher *ReceiveMatcher) Match(actual interface{}) (success bool, err erro
 
 	var subMatcher omegaMatcher
 	var hasSubMatcher bool
-
-	if matcher.Arg != nil {
-		subMatcher, hasSubMatcher = (matcher.Arg).(omegaMatcher)
+	var resultReference interface{}
+
+	// Valid arg formats are as follows, always with optional POINTER before
+	// optional MATCHER:
+	//   - Receive()
+	//   - Receive(POINTER)
+	//   - Receive(MATCHER)
+	//   - Receive(POINTER, MATCHER)
+	args := matcher.Args
+	if len(args) > 0 {
+		arg := args[0]
+		_, isSubMatcher := arg.(omegaMatcher)
+		if !isSubMatcher && reflect.ValueOf(arg).Kind() == reflect.Ptr {
+			// Consume optional POINTER arg first, if it ain't no matcher ;)
+			resultReference = arg
+			args = args[1:]
+		}
+	}
+	if len(args) > 0 {
+		arg := args[0]
+		subMatcher, hasSubMatcher = arg.(omegaMatcher)
```

Comment:
The previous implementation took the first arg, if there are one or more args, silently ignoring additional unused args.

In turn, if there was a _at least_ a first arg (`matcher.Arg != nil`) it differentiated between POINTER or MATCHER. This user-facing behavior is preserved, the corresponding unit tests are unmodified in this respect.

The change now accepts up to two optional args, where in case of two args the order must be POINTER, then MATCHER.

This should be only a breaking change to unit tests that in the past accidentally passed two or more args, with all args except for the first being silently ignored. This undefined behavior is now caught and reported.

Is there a different breaking change that I'm missing here?

Reply:
On second thought, maybe `Receive` was being used incorrectly but calling `Receive(nil)` with `nil` as the first argument worked before and [it threw an error on upgrading](https://github.com/onsi/gomega/pull/750/files#diff-9bcdd3004ba6d4c81b6791fcedb7588333eabda417284d3d38805166a5b6ee81R57). It's more explicit now anyway though, so not actually a problem.

Reply:
Ah, I wasn't aware of the explicit `nil` usage.